File: tools/ghidra_scripts/DecompileAt.py
# DecompileAt.py
# @category mn103

from ghidra.app.decompiler import DecompInterface


def parse_addr(value):
    v = value.strip().lower()
    if v.startswith("0x"):
        return int(v[2:], 16)
    return int(v, 10)


target = toAddr(0x2000)
args = getScriptArgs()
if args and len(args) > 0 and args[0].strip():
    target = toAddr(parse_addr(args[0]))

func = getFunctionContaining(target)
if func is None:
    disassemble(target)
    createFunction(target, None)
    func = getFunctionContaining(target)

if func is None:
    print("No function found at %s" % target)
    exit()

iface = DecompInterface()
iface.openProgram(currentProgram)

result = iface.decompileFunction(func, 60, monitor)
print("Function: %s @ %s" % (func.getName(), func.getEntryPoint()))
print("Completed: %s" % result.decompileCompleted())
print("TimedOut: %s" % result.isTimedOut())
err = result.getErrorMessage()
print("Error: %s" % (err if err else ""))

if result.decompileCompleted():
    dfunc = result.getDecompiledFunction()
    if dfunc is not None:
        print("--- Decompiled C ---")
        print(dfunc.getC())

File: tools/ghidra_scripts/DecompileAt.java
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Function;

public class DecompileAt extends GhidraScript {
    @Override
    public void run() throws Exception {
        Address target = currentAddress;
        String[] args = getScriptArgs();
        if (args != null && args.length > 0 && !args[0].isEmpty()) {
            target = toAddr(args[0]);
        } else if (target == null) {
            target = currentProgram.getMinAddress();
        }

        Function fn = currentProgram.getFunctionManager().getFunctionContaining(target);
        if (fn == null) {
            disassemble(target);
            createFunction(target, null);
            fn = currentProgram.getFunctionManager().getFunctionContaining(target);
        }
        if (fn == null) {
            println("No function at " + target);
            return;
        }

        DecompInterface ifc = new DecompInterface();
        ifc.openProgram(currentProgram);
        DecompileResults res = ifc.decompileFunction(fn, 30, monitor);

        println("Function: " + fn.getName() + " @ " + fn.getEntryPoint());
        println("Completed: " + res.decompileCompleted());
        println("TimedOut:  " + res.isTimedOut());
        println("Error:     " + res.getErrorMessage());
        if (res.getDecompiledFunction() != null) {
            println("---- Decompiled C ----");
            println(res.getDecompiledFunction().getC());
        }
    }
}
